common/oauthTokenManager.go

```diff
@@ -713,16 +724,134 @@ func (credInfo *OAuthTokenInfo) GetNewTokenFromMSI(ctx context.Context) (*adal.T
+// For details, please refer to https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/overview
+// Note: Currently the msiTokenHTTPClient timeout is configured for 30 secs. Should be reduced to 5 sec as IMDS endpoint is local to the machine.
+// Without this change, if some router is configured to not return "ICMP unreachable" then it will take 30 secs to timeout and fallback to ARC.
+// For now, this has been mitigated by checking Arc first, and then Azure
+func (credInfo *OAuthTokenInfo) GetNewTokenFromMSI(ctx context.Context) (*adal.Token, error) {
+	// Try Arc VM
+	req, resp, err := credInfo.queryIMDS(MSIEndpointArcVM, Resource, IMDSAPIVersionArcVM, ctx)
 	if err != nil {
-		return nil, fmt.Errorf("please check whether MSI is enabled on this PC, to enable MSI please refer to https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/qs-configure-portal-windows-vm#enable-system-assigned-identity-on-an-existing-vm. (Error details: %v)", err)
+		// Try Azure VM since there was an error in trying Arc VM
+		reqAzureVM, respAzureVM, errAzureVM := credInfo.queryIMDS(MSIEndpointAzureVM, Resource, IMDSAPIVersionAzureVM, ctx)
+		var serr syscall.Errno
+		errorConverted := errors.As(err, &serr)
+		if errAzureVM != nil {
+			if errorConverted {
+				var econnrefusedValue int
+				if runtime.GOOS == "linux" {
+					econnrefusedValue = int(syscall.ECONNREFUSED)
+				} else if runtime.GOOS == "windows" {
+					econnrefusedValue = WSAECONNREFUSED
+				} else {
+					econnrefusedValue = -1
+				}
+				if int(serr) == econnrefusedValue {
+					// If connection to Arc endpoint was refused
+					return nil, fmt.Errorf("please check whether MSI is enabled on this PC, to enable MSI please refer to https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/qs-configure-portal-windows-vm#enable-system-assigned-identity-on-an-existing-vm: %v", errAzureVM)
+				} else if checkIfWWWAuthenticateUnavailable(resp) {
+					// If response was malformed
+					return nil, fmt.Errorf("unknown process running at the endpoint: %v", err)
+				} else {
+					// If there is some other issue
+					return nil, fmt.Errorf("failed to process request at MSI endpoint: %v", err)
+				}
+			} else if checkIfWWWAuthenticateUnavailable(resp) {
+				// If response was malformed
+				return nil, fmt.Errorf("unknown process running at the endpoint: %v", err)
+			} else {
+				// If there is some other issue
+				return nil, fmt.Errorf("failed to process request at MSI endpoint: %v", err)
+			}
+		} else {
+			req, resp = reqAzureVM, respAzureVM
+		}
+	} else if checkIfWWWAuthenticateUnavailable(resp) {
+		reqAzureVM, respAzureVM, errAzureVM := credInfo.queryIMDS(MSIEndpointAzureVM, Resource, IMDSAPIVersionAzureVM, ctx)
```

Comment:
write a comment above this

// Not valid response from ARC IMDS endpoint. Some other process listening on it?